src/element.toit

```diff
@@ -0,0 +1,530 @@
+        label_ = value
+        left_ = null  // Trigger recalculation.
+        return
+    invalidate
+    label_ = value
+    left_ = null  // Trigger recalculation.
+    invalidate
+
+  orientation= value/int -> none:
+    if value == orientation_: return
+    invalidate
+    orientation_ = value
+    left_ = null  // Trigger recalculation.
+    invalidate
+
+  alignment= value/int -> none:
+    if value == alignment_: return
+    invalidate
+    alignment_ = value
+    left_ = null  // Trigger recalculation.
+    invalidate
+
+  draw canvas /Canvas -> none:
+    x := x_
+    y := y_
+    if not (x and y): return
+    if alignment_ != ALIGN_LEFT:
+      text_width := font_.pixel_width label_
+      if alignment_ == ALIGN_CENTER: text_width >>= 1
+      if orientation_ == ORIENTATION_0:
+        x -= text_width
+      else if orientation_ == ORIENTATION_90:
+        y += text_width
+      else if orientation_ == ORIENTATION_180:
+        x += text_width
+      else:
+        assert: orientation_ == ORIENTATION_270
+        y -= text_width
+    canvas.text x y --text=label_ --color=color_ --font=font_ --orientation=orientation_
+
+/**
+A superclass for elements that can draw themselves.  Override the
+  $draw method in your subclass to draw on the canvas.  The $w
+  and $h methods/fields are used to determine the size of the element
+  for redrawing purposes.
+
+Drawing operations are automatically clipped to w and h.
+*/
+abstract class CustomElement extends ClippingDiv:
+  abstract w -> int?
+  abstract h -> int?
+
+  constructor --x/int?=null --y/int?=null --w/int?=null --h/int?=null:
+    super --x=x --y=y --w=w --h=h
+
+  invalidate:
+    if change_tracker and x and y and w and h:
+      change_tracker.child_invalidated_element x y w h
+
+  draw canvas/Canvas -> none:
+    if not (x and y): return
+    analysis := canvas.bounds_analysis x y w h
+    if analysis == Canvas.DISJOINT: return
+    autocropped := analysis == Canvas.CANVAS_IN_AREA or analysis == Canvas.COINCIDENT
+    old_transform := canvas.transform
+    canvas.transform = old_transform.translate x_ y_
+    Background.draw background_ canvas 0 0 w h --autocropped=autocropped
+    custom_draw canvas
+    if border_: border_.draw canvas 0 0 w h
+    canvas.transform = old_transform
+
+  /**
+  Override this to draw your custom element.  The coordinate system is
+    the coordinate system of your element, ie the top left is 0, 0.
+  The background has already been drawn when this is called, and the
+    frame will be drawn afterwards.
+  */
+  abstract custom_draw canvas/Canvas -> none
+
+/**
+A ClippingDiv is like a div, but it clips any draws inside of it.  It can
+  have a shadow or other drawing outside its raw x y w h area, called the
+  decoration.
+For style purposes it has the type "div", not "clipping-div".
+Because it has clipping and compositing, it can have more interesting borders
+  like rounded corners.
+*/
+class ClippingDiv extends Div:
```

Comment:
Could `--clipping` be just an argument to the constructor of the `Div` class?
Internally it could still return a `ClippingDiv_`, but the user wouldn't need to deal with different classes (especially since their name is "div" in both cases).
Could `clipping` just be a boolean attribute that can be changed? (Does that even make sense?).